arch/arm/core/userspace.S

```diff
@@ -0,0 +1,154 @@
+/*
+ * Userspace and service handler hooks
+ *
+ * Copyright (c) 2017 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+#include <offsets_short.h>
+#include <toolchain.h>
+#include <linker/sections.h>
+#include <kernel_structs.h>
+#include <arch/cpu.h>
+#include <syscall.h>
+
+_ASM_FILE_PROLOGUE
+
+GTEXT(_arm_userspace_enter)
+GTEXT(_arm_do_syscall)
+GDATA(_kernel)
+
+/* Imports */
+GTEXT(_k_syscall_table)
+
+/**
+ *
+ * User space entry function
+ *
+ * This function is the entry point to user mode from privileged execution.
+ * The conversion is one way, and threads which transition to user mode do
+ * not transition back later, unless they are doing system calls.
+ *
+ */
+SECTION_FUNC(TEXT,_arm_userspace_enter)
+    /* preserve arguments */
+    mov lr,r0
+    push {r1,r2,r3,lr}
+
+    /* disable mpu stack guard to keep us from triggering guard region */
+    bl arm_core_mpu_disable
+    pop {r1,r2,r3,lr}
+
+    /*
+     * calculate end of stack and set SP to it
+     */
+    ldr r0, [sp]
+    ldr ip, [sp,#4]
+    add r0,ip
+
+    msr PSP, r0
+
+    /* push args to stack, pad end with 2 words of 0x0 */
+    sub sp,8
+    push {r1,r2,r3,lr}
+    ldr r3,=0
+    str r3,[sp, #20]
+    str r3,[sp, #24]
+
+    /* clear the stack area to clean out privileged data */
+#ifdef CONFIG_INIT_STACKS
+    ldr r1,=0xaaaaaaaa
+#else
+    ldr r1,=0
+#endif
+    mov r2,ip
+    sub r2,#24		/* don't erase the pushed regs */
+    bl memset
+
+#ifdef CONFIG_MPU_STACK_GUARD
+    /* setup arguments to thread_entry function */
+    ldr r0,=_kernel
+    ldr r0,[r0, #_kernel_offset_to_current]
+    bl configure_mpu_stack_guard
+    pop {r0,r1,r2,lr}
+#endif
+
+    /* change processor mode to unprivileged */
+    mov ip, #3
+    msr CONTROL, ip
+
+    /* return to thread_entry */
```

Comment:
i don't think this is implemented correctly.

it looks like this instruction will directly enter the user_entry function.
This is incorrect. What we have to do is setup the stack/registers so that we enter _thread_entry() with user_entry and its three arguments as parameters to *that* function.
_thread_entry never returns, and calls k_thread_abort() and does other housekeeping if user_entry returns

Reply:
bad comment.  This returns back to the point of the svc.  This does bring up the question of how we control who makes the call to svc.

Reply:
I think what Andrew mentioned is that the better implementation should be like the following:

_arm_userspace_enter(k_thread_entry_t user_entry,
					       void *p1, void *p2, void *p3,
					       u32_t stack_end,
					       u32_t stack_start)
{
    .....
    .....
    _thread_entry(user_entry, p1, p2, p3);
}

but the current implementation is:

_arm_userspace_enter(k_thread_entry_t user_entry,
					       void *p1, void *p2, void *p3,
					       u32_t stack_end,
					       u32_t stack_start)
{
    .....
    .....
    user_entry(p1, p2, p3);
}